File: cpp/protein-translation/protein_translation.cpp
#include "protein_translation.h"

namespace protein_translation {
    vector<string> proteins(const string &rna) {
        vector<string> results;
        for (int i = 0; i <= static_cast<int>(rna.length()) - 3; i += 3) {
            const string seq = rna.substr(i, 3);
            if (seq == "AUG") results.emplace_back("Methionine");
            if (seq == "UUU" || seq == "UUC") results.emplace_back("Phenylalanine");
            if (seq == "UUA" || seq == "UUG") results.emplace_back("Leucine");
            if (seq == "UCU" || seq == "UCC" || seq == "UCA" || seq == "UCG") results.emplace_back("Serine");
            if (seq == "UAU" || seq == "UAC") results.emplace_back("Tyrosine");
            if (seq == "UGU" || seq == "UGC") results.emplace_back("Cysteine");
            if (seq == "UGG") results.emplace_back("Tryptophan");
            if (seq == "UAA" || seq == "UAG" || seq == "UGA") break;

        }
        return results;
    }
}

File: elixir/protein-translation/lib/protein_translation.ex
defmodule ProteinTranslation do
  def of_rna(rna) do
    rna
    |> String.graphemes
    |> Enum.chunk_every(3)
    |> Enum.map(&List.to_string/1)
    |> Enum.reduce_while([], fn codon, acc ->
      case of_codon(codon) do
        {:ok, "STOP"} -> {:halt, acc}
        {:ok, protein} -> {:cont, [protein | acc]}
        {:error, _} = error -> {:halt, error}
      end
    end)
    |> case do
      {:error, _} -> {:error, "invalid RNA"}
      proteins -> {:ok, Enum.reverse(proteins)}
    end
  end

  def of_codon(codon) when codon in ["UGU", "UGC"], do: {:ok, "Cysteine"}
  def of_codon(codon) when codon in ["UUA", "UUG"], do: {:ok, "Leucine"}
  def of_codon(codon) when codon in ["UUU", "UUC"], do: {:ok, "Phenylalanine"}
  def of_codon(codon) when codon in ["UAU", "UAC"], do: {:ok, "Tyrosine"}
  def of_codon(codon) when codon in ["UCU", "UCC", "UCA", "UCG"], do: {:ok, "Serine"}
  def of_codon(codon) when codon in ["UAA", "UAG", "UGA"], do: {:ok, "STOP"}
  def of_codon("AUG"), do: {:ok, "Methionine"}
  def of_codon("UGG"), do: {:ok, "Tryptophan"}
  def of_codon(_), do: {:error, "invalid codon"}
end

